### ISSUE TITLE
Document `./write`

### ISSUE THREAD

**Issue by tychota** [lightvector/KataGo#5]
Once again, nice works :)

The write command is undocumented. I think it is usefull to start "master" like bots:
I assume its usage would be something like:

`./write -pool-size 50000 -train-shards 10 -val-game-prob 0.05 -gamesdir <DIRECTORY WITH SGF FILES> -output <TARGET>.h5`

Once we get the target h5, how to include it in the selfpay loop ?

I will be happy to open a PR with those information.

**Comment by lightvector** [comment]
Thanks for the interest!

The write command is left over from https://github.com/lightvector/GoNN and you can see the usage there. It was used for supervised learning from SGF game records along with the python scripts in GoNN that in this repo have since been moved to `../python/deprecated/`

I have not touched or tested the `write` command in a long time. Likely it still works, and I've tried to make it at least compile as I've changed things, although there is a chance it is broken by some intermediate changes since the tip of GoNN (since some of the changes since then also touched some fundamental things like the board history representation and neural net input feature construction, which write also uses).

The current (non-deprecated) self-play training scripts do not use h5 files, so as far as I know, the two ways of training neural nets (supervised vs selfplay) are incompatible right now. I've left the write executable in there in case, as you said, in the future I want to work on it to make it work with the current code, but it would take some development effort.

**Comment by tychota** [comment]
Interesting. So kata-bot on ogs is full "zero" ?

I would be interested to invest supervised approach (then making the networks stronger with selfplay) so I think I will have a look at the code to see were I can put games.

Maybe I can juste load h5 files, save as numpy in the right directory and fake the training part as if human games were selfplayed.

> I've generally not been happy with .h5 as compared to the new format used by selfplay using .npz files, so if I were to revive supervised training, I would likely drop h5 usage entirely and make it also use numpy as well. 

Makes sense.

**Comment by lightvector** [comment]
Yep, kata-bot is full zero. I think using write will be a very nontrivial amount of work, but if you're motivated to do it, let me know if you have any questions. If you'd like to see the current training data format, `cpp/dataio/numpywrite.*` is the low-level code for generating a zipped file of numpy data, which is then used by `cpp/dataio/trainingwrite.*` to write out the self-play data. In `cpp/dataio/trainingwrite.h` in the definition of `struct TrainingWriteBuffers` you can see some comments documenting about what every channel is. A bunch of the channels are not used by the training, they are merely for logging or informational purposes (for example the ones embedding the hash that allows matching up the training row with the sgf file).

I plan to be releasing the training data and full models history from KataGo's main training some time in the next week if possible (once I get large-amounts-of-data-hosting and other logistics solved), so you can see what the actual training data looks like. Attached is the readme that I will be including with that data, which you can also look at for documentation on the data format. 
[README.txt](https://github.com/lightvector/KataGo/files/2922336/README.txt)

**Comment by lightvector** [comment]
Just revisiting this issue. Note that KataGo's full run history training data is available here: https://d3dndmfyhecmj0.cloudfront.net/

I've actually pondered the idea myself of doing something much gentler and closer to 'zero' but still interesting - use human game positions to seed the starting points for games for self-play training, so that training gets to see and better learn a great variety of human joseki that it wouldn't normally play on its own. And who knows - maybe it will like some of them and play them more.

Anyways, let me know if you're still interested, happy to chat more. Incorporating human games for a 'master' version is still going to be a nontrivial work, probably not doable without some more coding, but could be possible.

**Comment by tychota** [comment]
Super interested to help but not sure about the time I can spend next month.

I think Leela Master does also a split training (the other way round, humans are better at gaining points than leela in the endgame).
Not sure how but maybe using joseki database for like 5% of games can be helpful so it learns old joseki.

Would that be another run? If so can we try to build a distributed one?

**Comment by lightvector** [comment]
For me, getting a stronger run and getting better support for more platforms, OpenCL, precompiled binaries, etc takes high priority over this, so I will not be working on it for quite a while, if ever.

However, I'm happy to chat and help troubleshoot if you're interesting in trying it, although it will have to involve getting your hands dirty making a branch and hacking on the code. Like Leela Zero I think I would not try to incorporate it directly into a run, but do it outside of the main selfplay loop. The steps would be to un-deprecate write and fix it to be able to write data in the new format (python npz files). Then some modifications to the shuffle script to make it shuffle that data along with the existing downloadable training data that I linked from the full run above into TF record files. If that was all working, then you could run a training for a few days on your own GPU.

Maybe best to wait for a new ongoing (unreleased) run of KataGo to finish though. Some stronger nets will be coming out soon. And also the training data format for KataGo has changed a little as of the tip of master compared to the released training data (so if you wanted to work on things right away, you would have to do it from closer to the old tagged revision in the release, prior to the training data format change).

**Comment by tychota** [comment]
All clear.

> Maybe best to wait for a new ongoing (unreleased) run of KataGo to finish though. Some stronger nets will be coming out soon.

🥰

> For me, getting a stronger run and getting better support for more platforms, OpenCL, precompiled binaries, etc takes high priority over this, so I will not be working on it for quite a while, if ever.

For sure.

Out of topic but better integration with lizzie will be nice (like not having a warning at start about leela `Lizzie requires version 0.15 or later of Leela Zero for analysis (found 1.0)`, territory estimation, and stuff that comes from GoNN like maybe "ladder" mode).

But I guess that's a point for another issue.

> However, I'm happy to chat and help troubleshoot if you're interested in trying it, although it will have to involve getting your hands dirty making a branch and hacking on the code. Like Leela Zero I think I would not try to incorporate it directly into a run, but do it outside of the main self play loop. The steps would be to un-deprecate write and fix it to be able to write data in the new format (python npz files). 


I will try that in two or three weeks.

**Comment by lightvector** [comment]
Hehe. That's a hardcoded thing in Lizzie where it assumes that the bot is
Leela Zero when it tries to parse the version. You can pass in
-override-version 0.16 or 0.17 to make KataGo lie about what version it is
and report the number that Leela Zero would report, so that Lizzie will
think that it is interacting with the right version of Leela Zero.

Fixing it probably is more of a "submit a pull request to fix Lizzie" task
than anything that requires a code change in KataGo, but I haven't poked
into Lizzie source yet.

**Comment by tychota** [comment]
https://github.com/featurecat/lizzie/blob/d0e7040379e1a23487e4e1bcfd1447bd676a6339/src/main/java/featurecat/lizzie/analysis/Leelaz.java#L315-L325

**Comment by lightvector** [comment]
Closing this. Lizzie and KataGo work well together as of latest versions, and "write" continues to remain deprecated and do-not-use, so I'm not going to bother documenting. If there's something further to do here, let me know. :)
